### PR TITLE
py: implemented rounding floats to precision

### PR DESCRIPTION
**Pull request by turbinenreiter**
in continuation of https://github.com/micropython/micropython/pull/1376
- supports rounding floats to precision
- uses macros
- added warning may cause overflow

### CODE REVIEW

**Comment by turbinenreiter** [comment]
oh man, now these macros break the stmhal builds.
I need to figure out how to get Travis running on my own repo, so I see these results before creating noisy pull requests.

**Comment by danicampora** [comment]
It's really easy, I don't remember the exact steps, but is like 2 or 3
clicks on your repo config.

**Comment by pfalcon** [comment]
Well, unfortunately, there's swing of pendulum in another direction comparing to #1376 ;-). Let's start with the main issue: there *should* be tests. Just, as comments to the above issue said, you should avoid using N > 5 in the test.

File: py/modbuiltins.c [hunk 1]
```diff
@@ -440,8 +440,18 @@ STATIC mp_obj_t mp_builtin_round(mp_uint_t n_args, const mp_obj_t *args) {
     mp_int_t num_dig = 0;
     if (n_args > 1) {
         num_dig = mp_obj_get_int(args[1]);
-        if (num_dig > 0) {
-            mp_not_implemented("round(..., N>0)");
+        if (num_dig > 0 && num_dig < 7) {
```
Comment (pfalcon) [inline py/modbuiltins.c]: There should not be such check ("< 7") here, it's implementation-specific.

File: py/modbuiltins.c [hunk 2]
```diff
@@ -440,8 +440,18 @@ STATIC mp_obj_t mp_builtin_round(mp_uint_t n_args, const mp_obj_t *args) {
...
+            mp_float_t val = mp_obj_get_float(o_in);
+            mp_float_t mult = MICROPY_FLOAT_C_FUN(pow)(10, num_dig);
+            mp_float_t rounded = MICROPY_FLOAT_C_FUN(round)(val * mult) / mult;
+            return mp_obj_new_float(rounded);
+        } else if (num_dig > 0 && num_dig > 6) {
+            mp_not_implemented("round(..., N>6)");
+        } else if (num_dig < 0) {
+            mp_float_t val = mp_obj_get_float(o_in);
+            mp_float_t mult = MICROPY_FLOAT_C_FUN(pow)(10, MICROPY_FLOAT_C_FUN(fabs)(num_dig));
+            mp_float_t rounded = MICROPY_FLOAT_C_FUN(round)(val * mult);
```
Comment (pfalcon) [inline py/modbuiltins.c]: I don't think this implementation is correct. I'm still -0.3 on not implementing < 0 case at all (how often it is used? ~never).

File: py/modbuiltins.c [hunk 3]
```diff
@@ -431,7 +431,7 @@ STATIC mp_obj_t mp_builtin_repr(mp_obj_t o_in) {
 MP_DEFINE_CONST_FUN_OBJ_1(mp_builtin_repr_obj, mp_builtin_repr);
 
 STATIC mp_obj_t mp_builtin_round(mp_uint_t n_args, const mp_obj_t *args) {
-    // TODO really support second arg
+    // TODO second arg may lead to overflow
```
Comment (pfalcon) [inline py/modbuiltins.c]: I suggest adding "overflow" comment directly above the line where overflow may happen.

**Comment by pfalcon** [comment]
> how to get Travis running on my own repo

As @danicampora points, it's easy, but you don't even need to - we have (well, github has) Travis running on pull requests exactly to let people skip running Travis on their own repos.

File: py/modbuiltins.c (hunk 1 above)
Comment (turbinenreiter) [inline py/modbuiltins.c]: You mean leave it out entirely?

File: py/modbuiltins.c (hunk 2 above)
Comment (turbinenreiter) [inline py/modbuiltins.c]: will remove it

Comment (danicampora) [inline py/modbuiltins.c]: I agree with @pfalcon.

**Comment by turbinenreiter** [comment]
@pfalcon right, I'm going to add tests.

Should I make changes in new commits, or should I squash them into one?

File: py/modbuiltins.c (hunk 2 above)
Comment (pfalcon) [inline py/modbuiltins.c]: > will remove it

Well, so far it was just my own vote (well, @danicampora now agrees too). Feel free to wait for other reviewers. Ideal solution would be to see how much code size <0 case adds and then decide.

**Comment by pfalcon** [comment]
> Should I make changes in new commits, or should I squash them into one?

I suggest adding new commits, then when you get green light, squash everything to be merged.

File: py/modbuiltins.c (hunk 1 above)
Comment (pfalcon) [inline py/modbuiltins.c]: Yes, no need to test for < 7. Actually, maybe no need to test for >0, then <0 case should actually automagically work.

**Comment by turbinenreiter** [comment]
ok, see qemu-arm test fails:
```
basics/builtin_round.py: Traceback (most recent call last):
  File "<stdin>", line 6
SyntaxError: decimal numbers not supported

  FAIL test_main.c:42: Uncaught exception
  [builtin_round.py FAILED]
```
I guess I'd need separate tests for when floats are supported and when not? Not sure how to do that.

**Comment by danicampora** [comment]
Make two tests, one with floats and another with ints. Place the float one in `tests/float` and it should be fine.

**Comment by pfalcon** [comment]
> Make two tests, one with floats and another with ints. Place the float one in tests/float and it should be fine.

Or more specifically: leave basics/builtin_round.py as is, add float/builtin_round_float.py with new tests.

**Comment by turbinenreiter** [comment]
opinions?

**Comment by pfalcon** [comment]
> opinions?

This is common theme - github doesn't send notification on new version of PR being pushed, so pushing your changes is not replacement for a bit of human chit-chat ;-).

From a quick look, looks good, will target to merge later today.

**Comment by turbinenreiter** [comment]
Realized that because of the silence ;)

**Comment by pfalcon** [comment]
Perfect, thanks! Merged.

**Pull request closed without merging**